### PR TITLE
Fix bug when adding tier price to simple product from product edit page.

### PR DESCRIPTION
**Pull request by markbeazley**
Discovered a bug when adding/editing tier prices on a simple product as the form has a different root object Product vs Product Variant, I've managed to work around this, by checking if the form root is of type ProductInterface and if so grab the first ProductVariant and use that.

### CODE REVIEW

**Comment by markbeazley** [comment]
Should I open an issue for this or would it be possible to get this considered for merging without an issue being created?

**Comment by mamazu** [comment]
@kortwotze will have a look at it as he is currently using this in our production environment.

**Comment by kortwotze** [comment]
Can you elaborate a bit more on what happened when you edited a simple product?

File: src/Validator/TierPriceUniqueValidator.php [hunk 1]
```diff
@@ -59,7 +61,15 @@ public function validate($value, Constraint $constraint): void
             );
         }
 
-        $otherTierPrices = $this->context->getRoot()->getData()->getTierPrices();
+        $formData = $this->context->getRoot()->getData();
```
Comment (kortwotze) [inline src/Validator/TierPriceUniqueValidator.php]: I'd suggest naming this `$productVariant` for clarity.

**Comment by markbeazley** [comment]
Basically when you edit a tier price from the product edit route (rather than the variant edit route) it would try and call getTierPrices() on the Product rather than the variant due to the way the forms are nested.

**Comment by kortwotze** [comment]
Mh - I can't reproduce it on my local sylius. Adding a tierprice on `admin/products/11/edit` is no problem here. Gets persisted in database and shown after reload.
Not trying to criticize here, just trying to comprehend properly. :)

**Comment by kortwotze** [comment]
Disclaimer: I am using 1.5.0. where this file is not even part of. Okay, I take back of most that I said and try to understand what @mamazu has merged in 1.6.*.

**Comment by markbeazley** [comment]
Yeah it is a bug in part of the validator I added when I added the customer groups features as annoyingly the built in UniqueEntity validator doesn't catch cases where a user tries to add two identical tier prices at the same time (it only checks against entities that were in the database at the time the form was submitted). So I wrote something that gets the submitted data and checks uniqueness in that data.

File: src/Validator/TierPriceUniqueValidator.php (hunk 1 above)
Comment (markbeazley) [reply src/Validator/TierPriceUniqueValidator.php]: I can change it to that, however there is a possibility it is a Product, and on the next line I call `$formData->hasVariants()` and `$productVariant->hasVariants()` seems a little odd.

**Comment by kortwotze** [comment]
Right - thanks for the insight. Now I am just wondering why the form can be rendered with a product as a root, rather than a variant (which should be the case all the time from what I see).

I still consider this fix valid though, just waiting for the full clarification and then we merge it. No issue needed here, I think :)

**Comment by markbeazley** [comment]
Yeah I was a little puzzled by that myself, Not sure if I should have been using something else other than getRoot().

**Comment by pierre-H** [comment]
same problem ...

File: src/Validator/TierPriceUniqueValidator.php [hunk 2]
```diff
@@ -59,7 +61,15 @@ public function validate($value, Constraint $constraint): void
...
+        if ($formData instanceof ProductInterface && $formData->hasVariants()) {
+            $formData = $formData->getVariants()->first();
```
Comment (kortwotze) [inline src/Validator/TierPriceUniqueValidator.php]: Maybe also check if its exactly 1 variant on the product, not just using the first.

File: src/Validator/TierPriceUniqueValidator.php (hunk 1 above)
Comment (kortwotze) [reply src/Validator/TierPriceUniqueValidator.php]: I see.

File: src/Validator/TierPriceUniqueValidator.php (hunk 2 above)
Comment (markbeazley) [reply src/Validator/TierPriceUniqueValidator.php]: So change the if to this:

`if ($formData instanceof ProductInterface && count($formData->getVariants()) === 1)`

Comment (mamazu) [reply src/Validator/TierPriceUniqueValidator.php]: Looks good to me.

Comment (markbeazley) [reply src/Validator/TierPriceUniqueValidator.php]: I've made the change as mentioned above.

Comment (peterukena) [reply src/Validator/TierPriceUniqueValidator.php]: :+1:

Comment (mamazu) [reply src/Validator/TierPriceUniqueValidator.php]: Should we merge it?

**Pull request merged**